astropy/io/ascii/tests/test_c_reader.py

```diff
@@ -886,3 +886,41 @@ def test_fast_tab_with_names(parallel, read_tab):
     head = ['A{0}'.format(i) for i in range(28)]
     table = read_tab(content, data_start=1,
                      parallel=parallel, names=head)
+
+# catch Windows environment since we cannot use _read() with custom fast_reader
+@pytest.mark.parametrize("parallel", [
+    pytest.mark.xfail(os.name == 'nt', reason=
+                      "Multiprocessing is currently unsupported on Windows")(True),
+    False])
+def test_fortran_reader(parallel):
+    """
+    Make sure that ascii.read() can read Fortran-style exponential notation
+    using the fast_reader.
+    """
+    text = 'A B C\n100.01D+99 2.0 3\n4.2d-1 5.0D-1 0.6D4'
+    expected = Table([[1.0001e101, 0.42], [2, 0.5], [3.0, 6000]], 
+                     names=('A', 'B', 'C'))
+
+    # C strtod (not-fast converter) can't handle Fortran exp
+    with pytest.raises(FastOptionsError) as e:
+        ascii.read(text, format='basic', guess=False, fast_reader=
+                   {'parallel': parallel, 'use_fast_converter': False,
+                    'exponent_style': 'D'})
+    assert 'fast_reader: exponent_style requires use_fast_converter' in str(e)
+
+    # Enable multiprocessing and the fast converter
+    table = ascii.read(text, format='basic', guess=False, 
+                       fast_reader={'parallel': parallel, 'exponent_style': 'D'})
+    assert_table_equal(table, expected)
+
+    text = 'A B C\n0.10001Q102 20.0Q-1 3\n.42q0 0.5 6.Q3'
+    table = ascii.read(text, format='basic', guess=False, 
+                       fast_reader={'parallel': parallel, 'exponent_style': 'q'})
+    assert_table_equal(table, expected)
+
+    # mixes and triple-exponents without any character using autodetect option
+    text = 'A B C\n1.0001+101 2.0E0 3\n.42d0 0.5 0.42-123'
```

Comment:
OK, that's probably something they won't (except for `1.23-100.2` I'd expect). More work on `tokenizer.c` to be done!